README.md
# Gematria (גימטריא) [![Build Status](https://travis-ci.org/adamzaninovich/gematria.png?branch=master)](https://travis-ci.org/adamzaninovich/gematria)

A Ruby gem that calculates Gematria. The gem comes with a few built-in correspondence tables, but allows for user-configurable tables. The gem supports raw conversion to number (by simple summation), mapping (breakdown of individual numbers), and reduction to a single digit (mispar katan mispari).

See the [website](http://adamzaninovich.github.com/gematria) for more detailed information.

The included correspondence tables are as follows:

![Hebrew Correspondence Table](http://adamzaninovich.github.com/gematria/images/hebrew.png)
![Greek Correspondence Table](http://adamzaninovich.github.com/gematria/images/greek.png)
![English Correspondence Table](http://adamzaninovich.github.com/gematria/images/english.png)
![English ALW Correspondence Table](http://adamzaninovich.github.com/gematria/images/alw.png)

## Documentation

Documentation at [rubydoc.info](http://rubydoc.info/github/adamzaninovich/gematria).

## Installation

Add this line to your application's Gemfile:

    gem 'gematria'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install gematria

## Usage

    name = Gematria::Calculator.new "Adam", :english
    name.converted          # => 46
    name.mapped.join(" + ") # => "1 + 4 + 1 + 40"
    name.reduced            # => 1

    gematria = Gematria::Calculator.new "Gematria is fun!", :english
    gematria.converted      # => 818
    gematria.mapped         # => [7, 5, 40, 1, 200, 90, 9, 1, 0, 9, 100, 0, 6, 300, 50, 0]

Setting a global table

    Gematria::Tables.set_table :hebrew
    alephbet = Gematria::Calculator.new 'אבגדהוזחטיכלמנסעפצקרשתךםןףץ'
    alephbet.converted == 4995 # => true

Defining a custom table

    Gematria::Tables.add_table :mini, 'a' => 1, 'b' => 10, 'c' => 100
    abcd = Gematria::Calculator.new 'abcd', :mini
    abcd.mapped    # => [1,10,100,0]
    abcd.converted # => 111
    abcd.reduced   # => 3

## Known Issues

* Not compatible with Ruby 1.8.7 or below. (If you need 1.8.7, use v0.0.2)

If you find any issues or errors with this gem, please report them on [Github Issues](http://github.com/adamzaninovich/gematria/issues).

## Contributing

Please submit documentation and specs with your feature.

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

Copyright (c) 2013 Adam Zaninovich

MIT License

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## v0.0.6

* Adds Greek and alternate English (ALW cipher) to built-in tables

## v0.0.5

* Improved documentation
* `Gematria::Tables` is now an instance of `Gematria::TableManager` rather than a method bucket

## v0.0.4

* Ability to set global table so as to not have to specify a table name with every new instance

## v0.0.3

* Removes support for 1.8.7 (due to encoding problems with Hebrew) (Now compatible with 1.9.2 - head)
* Adds Hebrew to built-in tables (now includes English and Hebrew)
* Adds support for custom tables
* Requires specifying a table when instantiating
* Changes api to use Gematria::Calculator (this may still not be stable)

## v0.0.2

* Ruby 1.8.7 compatible (Now compatible with 1.8.7 - head)

## v0.0.1

* initial release

spec/spec_helper.rb
# -*- encoding : utf-8 -*-
require 'rspec'
require 'gematria'


spec/gematria/tables_spec.rb
# -*- encoding : utf-8 -*-
require 'spec_helper'

module Gematria

	describe Tables do
		it "has a hash to store the tables" do
			expect(Tables::TABLES).to be_a Hash
		end

		describe "#set_table" do
			it "assigns a table to @current" do
				Tables.set_table(:english)
				expect(Tables.instance_variable_get :@current).to eq :english
			end
		end

		describe "#current" do
			context "when a table has been selected" do
				it "returns the previously selected table" do
					Tables.set_table(:hebrew)
					expect(Tables.current).to eq Tables.fetch(:hebrew)
				end
			end
			context "when a table has not been selected" do
				before do
					if Tables.instance_variable_defined? :@current
						Tables.instance_variable_set :@current, nil
					end
				end
				it "returns an empty hash" do
					expect(Tables.current).to eq({})
				end
			end
		end

		describe "#add_table" do
			context "when table is valid" do
				it "adds the table to the TABLES store" do
					Tables.add_table :valid, 'a'=>1,'b'=>2
					expect(Tables::TABLES).to include valid: {'a'=>1,'b'=>2}
				end
			end
			context "when table is invalid" do
				it "raises a TypeError" do
					expect {
						Tables.add_table :invalid, [['a',1],['b',2]]
					}.to raise_error TypeError
				end
			end
		end

		describe "#fetch" do
			context "when table exists" do
				it "returns the table" do
					Tables.add_table :mytable, 'a'=>1
					expect(Tables.fetch(:mytable)).to eq 'a'=>1
				end
			end
			context "when table does not exist" do
				it "returns an empty hash" do
					expect(Tables.fetch :non_existant_table).to eq({})
				end
			end
		end

		describe "#[]" do
			it "delegates to #fetch" do
				Tables.should_receive(:fetch).with(:key)
				Tables[:key]
			end
		end

	end

end

spec/gematria/english_spec.rb
require 'spec_helper'

describe Gematria::English do
  it "is instantiated with text and provides access to that text" do
    gematria = Gematria::English.new('mytext')
    expect(gematria.text).to eq 'mytext'
  end

  describe "#mapped" do
    it "returns an array of values for each character in text" do
      gematria = Gematria::English.new('abc')
      expect(gematria.mapped).to eq [1,2,3]
    end
  end

  describe "#converted" do
    it "returns the sum of the letter values" do
      gematria = Gematria::English.new('abc')
      expect(gematria.converted).to eq 6
    end
  end

  describe "#reduced" do
    context "when converted value is a single digit" do
      it "returns a single digit" do
        gematria = Gematria::English.new('abc')
        expect(gematria.reduced).to eq 6
      end
    end
    context "when converted value is more than one digit" do
      it "returns a single digit" do
        gematria = Gematria::English.new('abcdefghijklmnopqrstuvwxyz')
        expect(gematria.reduced).to eq 9 # 4095 -> 18 -> 9
      end
    end
  end
end

lib/gematria/tables.rb
# -*- encoding : utf-8 -*-
module Gematria
  module Tables
    # Stores available tables
    TABLES = {}

    # Sets the global current table
    #
    #     Tables.set_table(:english)
    #     Tables.current # => {'a'=>1,'b'=>2,...}
    #
    # @param table_name [Symbol] table name
    # @return [Symbol] table name
    def self.set_table(table_name)
      @current = table_name
    end

    # Gets the currently selected table
    #
    #     Tables.set_table(:english)
    #     Tables.current # => {'a'=>1,'b'=>2,...}
    #
    # @return [Hash{String => Number}]
    def self.current
      fetch(@current)
    end

    # Adds a table to the table store. A valid table must be a Hash with single character string keys and numerical values. 
    # A table that is not a Hash will raise a TypeError. The table name will be converted to a Symbol.
    #
    #     Tables.add_table :mini_english, 'a' => 1, 'b' => 2, 'c' => 3
    #
    # @param name [#to_sym] table name
    # @param table [Hash{String => Number}] table of characters pointing to corresponding values
    def self.add_table(name, table)
      if table.is_a? Hash
        TABLES[name.to_sym] = table
      else
        raise TypeError, 'Invalid table format' 
      end     
    end

    # Fetches specified table
    #
    #     Tables.fetch(:mini_english) # => {'a'=>1,'b'=>2,'c'=>3}
    #
    # @param key [Symbol] table name as a Symbol
    # @return [Hash{String => Number}]
    def self.fetch(key)
      TABLES.fetch(key, {})
    end

    # Fetches specified table (delegates to #fetch)
    #
    #     Tables[:mini_english] # => {'a'=>1,'b'=>2,'c'=>3}
    #
    # @param key [Symbol] table name as a Symbol
    # @return [Hash{String => Number}]
    def self.[](key)
      fetch(key)
    end

  end
end


lib/gematria/calculator.rb
# -*- encoding : utf-8 -*-
module Gematria
  # Performs the gematria calculations
  class Calculator
    
    attr_reader :text, :table

    # Takes the text to be converted and an optional table name
    # @param text [String] text to be converted
    # @param table_name [Symbol] optional table name to use in conversion. If none is specified, the global will be used (see Gematria::Tables#set_table) 
    def initialize(text, table_name=nil)
      @text = text
      if table_name.is_a? Symbol
        @table = Tables.fetch(table_name)
      else
        @table = Tables.current
      end
    end

    # Gets an array of the values for each character of text.
    # @example
    #   name = Gematria::English.new("Adam")
    #   name.mapped # => [1, 4, 1, 40]
    # @return [Array<Number>] numerical values attributed to each character of the original text
    def mapped
      text.each_char.map { |c| lookup_char c }
    end

    # Gets the summed gematria number of text.
    # @example
    #   name = Gematria::English.new("Adam")
    #   name.converted # => 46
    # This uses <tt>mapped</tt> internally and sums the results.
    # @return [Number] sum of the values of each character
    def converted
      mapped.inject(:+)
    end

    # Recursively reduces the <tt>converted</tt> value to a single digit mispar katan mispari style.
    # @example
    #   name = Gematria::English.new("Adam")
    #   name.reduced # => 1
    #   # "Adam" -> 46 -> 10 -> 1
    # @return [Number] single digit from reduction of <tt>converted</tt> value
    def reduced
      do_reduction_on converted
    end

    private
    
    def lookup_char(char)
      table.fetch(char.downcase, 0)
    end

    def do_reduction_on(number)
      if number < 10
        number
      else
        do_reduction_on number.to_s.each_char.map(&:to_i).inject(:+)
      end
    end

  end
end


lib/gematria/table_manager.rb
# -*- encoding : utf-8 -*-
module Gematria
  # Manages a set of tables that can be used to do the gematria conversion
  class TableManager

    attr_reader :tables

    def initialize
      @tables = {}
      @current = nil
    end

    # Sets the current table
    # @example
    #   Gematria::Tables.set_table(:english)
    # @param table_name [Symbol] table name
    # @return [Symbol] table name
    def set_table(table_name)
      @current = table_name
    end

    # Gets the currently selected table
    # @example After setting the table to <tt>:english</tt> with the <tt>#set_table</tt> method
    #   Gematria::Tables.current # => {'a'=>1,'b'=>2,...}
    # @return [Hash{String => Number}]
    def current
      fetch(@current)
    end

    # Adds a table to the table store. A valid table must be a Hash with single character string keys and numerical values. 
    # A table that is not a Hash will raise a TypeError. The table name will be converted to a Symbol.
    # @example
    #   Gematria::Tables.add_table :mini_english, 'a' => 1, 'b' => 2, 'c' => 3
    # @raise [TypeError] if table is not a valid Hash
    # @param name [#to_sym] table name
    # @param table [Hash{String => Number}] table of characters pointing to corresponding values
    # @return [void]
    def add_table(name, table)
      if table.is_a? Hash
        tables[name.to_sym] = table
      else
        raise TypeError, 'Invalid table format' 
      end     
    end

    # Fetches specified table
    # @example
    #   Gematria::Tables.fetch(:mini_english) # => {'a'=>1,'b'=>2,'c'=>3}
    # @param key [Symbol] table name as a Symbol
    # @return [Hash{String => Number}]
    def fetch(key)
      tables.fetch(key, {})
    end
    alias :[] :fetch

  end
end


lib/gematria/built_in_tables.rb
# -*- encoding : utf-8 -*-
module Gematria
  BUILT_IN_TABLES = {
    # applies "mispar hechrachi" method to English alphabet (http://www.inner.org/gematria/fourways.php)
    english: {
      'a' => 1,   'b' => 2,   'c' => 3,   'd' => 4,   'e' => 5,   'f' => 6,   'g' => 7,   'h' => 8,   'i' => 9,
      'j' => 10,  'k' => 20,  'l' => 30,  'm' => 40,  'n' => 50,  'o' => 60,  'p' => 70,  'q' => 80,  'r' => 90,
      's' => 100, 't' => 200, 'u' => 300, 'v' => 400, 'w' => 500, 'x' => 600, 'y' => 700, 'z' => 800
    },

    # English (ALW cipher)
    alw: {
      'a' => 1,   'l' => 2,   'w' => 3,   'h' => 4,   's' => 5,   'd' => 6,   'o' => 7,   'z' => 8,   'k' => 9, 
      'v' => 10,  'g' => 11,  'r' => 12,  'c' => 13,  'n' => 14,  'y' => 15,  'j' => 16,  'u' => 17,  'f' => 18, 
      'q' => 19,  'b' => 20,  'm' => 21,  'x' => 22,  'i' => 23,  't' => 24,  'e' => 25,  'p' => 26
    },

    # Mispar gadol table
    hebrew: {
      'א' => 1,   'ב' => 2,   'ג' => 3,   'ד' => 4,   'ה' => 5,   'ו' => 6,   'ז' => 7,   'ח' => 8,   'ט' => 9,
      'י' => 10,  'כ' => 20,  'ל' => 30,  'מ' => 40,  'נ' => 50,  'ס' => 60,  'ע' => 70,  'פ' => 80,  'צ' => 90,
      'ק' => 100, 'ר' => 200, 'ש' => 300, 'ת' => 400, 'ך' => 500, 'ם' => 600, 'ן' => 700, 'ף' => 800, 'ץ' => 900
    },

    # Greek (Isopsephy)
    greek: {
      'Α' => 1, 'Β' => 2,   'Γ' => 3,   'Δ' => 4,   'Ε' => 5,   'Ϝ' => 6,   'Ζ' => 7,   'Η' => 8,   'Θ' => 9, 
      'α' => 1, 'β' => 2,   'γ' => 3,   'δ' => 4,   'ε' => 5,   'ϛ' => 6,   'ζ' => 7,   'η' => 8,   'θ' => 9, 
      'Ι' => 10, 'Κ' => 20, 'Λ' => 30,  'Μ' => 40,  'Ν' => 50,  'Ξ' => 60,  'Ο' => 70,  'Π' => 80,  'Ϙ' => 90, 
      'ι' => 10, 'κ' => 20,  'λ' => 30,  'μ' => 40,  'ν' => 50,  'ξ' => 60,  'ο' => 70,  'π' => 80, 
      'Ρ' => 100,'Σ' => 200, 'Τ' => 300, 'Υ' => 400, 'Φ' => 500, 'Χ' => 600, 'Ψ' => 700, 'Ω' => 800,'ϡ' => 900, 
      'ρ' => 100,'σ' => 200, 'τ' => 300, 'υ' => 400, 'φ' => 500,'χ' => 600, 'ψ' => 700, 'ω' => 800
    }
  }
end

lib/gematria/english.rb
module Gematria
  class English < Struct.new(:text)
    # applies "mispar hechrachi" method to English alphabet (http://www.inner.org/gematria/fourways.php)
    CORRESPONDENCE_TABLE = {
      :a => 1,   :b => 2,   :c => 3,   :d => 4,   :e => 5,   :f => 6,   :g => 7,   :h => 8,   :i => 9,
      :j => 10,  :k => 20,  :l => 30,  :m => 40,  :n => 50,  :o => 60,  :p => 70,  :q => 80,  :r => 90,
      :s => 100, :t => 200, :u => 300, :v => 400, :w => 500, :x => 600, :y => 700, :z => 800
    }

    # Gets an array of the values for each character of text.
    #
    #   name = Gematria::English.new("Adam")
    #   name.mapped # => [1, 4, 1, 40]
    #
    # @return [Array<Number>] numerical values attributed to each character of the original text
    def mapped
      text.each_char.map { |c| lookup_char c }
    end

    # Gets the summed gematria number of text.
    #
    #   name = Gematria::English.new("Adam")
    #   name.converted # => 46
    #
    # This uses <tt>mapped</tt> internally and sums the results.
    # @return [Number] sum of the values of each character
    def converted
      mapped.inject(:+)
    end

    # Recursively reduces the <tt>converted</tt> value to a single digit mispar katan mispari style.
    #
    #   name = Gematria::English.new("Adam")
    #   name.reduced # => 1
    #
    # For example: "Adam" -> 46 -> 10 -> 1
    # @return [Number] single digit from reduction of <tt>converted</tt> value
    def reduced
      do_reduction_on converted
    end

    private
    def lookup_char(char)
      CORRESPONDENCE_TABLE.fetch(char.downcase.to_sym, 0)
    end

    def do_reduction_on(number)
      if number < 10
        number
      else
        do_reduction_on number.to_s.each_char.map(&:to_i).inject(:+)
      end
    end

  end
end